File: 2024/csharp/Puzzles/Dec02.cs
using csharp.utils;

namespace csharp.Puzzles;

public static class Dec02
{
    public static int solvePart1(string? date, bool useTestData = false)
    {
        Console.WriteLine("Day 02, Part 1: Safe reports\n");

        // create a data file reader and read the file.
        var dfr = new DataFileReader(date: date, useTestData: useTestData, part: 1);
        dfr.ReadFile();

        int totalSafeReports = 0;

        // Console.WriteLine($"Total number of reports: {dfr.Lines.Count}\n");

        // for each report (line)
        for (int i = 0; i < dfr.Lines.Count; i++) 
        {
            // split the report to get the levels
            string[] report = dfr.Lines[i].Split(" ", StringSplitOptions.RemoveEmptyEntries);

            if (reportIsSafe(report))
            {
                totalSafeReports++;
            }
        }

        // log the solution
        var outputString = useTestData ? "Total [using test data]" : "Total [using puzzle data]";
        Console.WriteLine($"{outputString}: {totalSafeReports}\n");

        return totalSafeReports;
    }

    public static int solvePart2(string date, bool useTestData = false)
    {
        // log current puzzle
        Console.WriteLine("Day 02, Part 2: Problem Dampener\n");

        // create a data file reader and read the file.
        var dfr = new DataFileReader(date: date, useTestData: useTestData, part: 1);
        dfr.ReadFile();

        int totalSafeReports = 0;

        // Console.WriteLine($"Total number of reports: {dfr.Lines.Count}\n");

        // for each report (line)
        for (int i = 0; i < dfr.Lines.Count; i++) 
        {
            // split the report to get the levels
            string[] report = dfr.Lines[i].Split(" ", StringSplitOptions.RemoveEmptyEntries);

            if (reportIsSafePt2(report))
            {
                totalSafeReports++;
            }
        }

        // log the solution
        var outputString = useTestData ? "Total [using test data]" : "Total [using puzzle data]";
        Console.WriteLine($"{outputString}: {totalSafeReports}");

        return totalSafeReports;
    }

    private static bool reportIsSafe(string[] report)
    {
        if (IsGraduallyIncreasing(report) || IsGraduallyDecreasing(report))
        {
            return true;
        }
        else
        {
            return false;
        }
    }
    
    private static bool reportIsSafePt2(string[] report)
    {
        // if the report is always increasing or decreasing and within tolerance, return true
        if (IsGraduallyIncreasing(report) || IsGraduallyDecreasing(report))
        {
            return true;
        }
        else if (safeWithBadValue(report))
        {
            return true;
        }
        else
        {
            return false;
        }
    }

    private static bool safeWithBadValue(string[] report)
    {
        for (int i = 0; i < report.Length; i++)
        {
            // try removing every value and see if the report is safe
            List<string> reportList = new List<string>(report);
            reportList.RemoveAt(i);
            var newReport = reportList.ToArray();
            if (IsGraduallyIncreasing(newReport) || IsGraduallyDecreasing(newReport))
            {
                return true;
            }
        }
        return false;
    }

    private static bool IsGraduallyIncreasing(string[] report)
    {
        for (int i = 0; i < report.Length - 1; i++)
        {
            var currentLevel = Convert.ToInt32(report[i]);
            var nextLevel = Convert.ToInt32(report[i+1]);
            if (currentLevel < nextLevel && IsWithinTolerance(currentLevel, nextLevel))
            {
                continue;
            }
            else
            {
                return false;
            }
        }
        return true;
    }

    private static bool IsGraduallyDecreasing(string[] report)
    {
        for (int i = 0; i < report.Length - 1; i++)
        {
            var currentLevel = Convert.ToInt32(report[i]);
            var nextLevel = Convert.ToInt32(report[i+1]);
            if (currentLevel > nextLevel && IsWithinTolerance(currentLevel, nextLevel))
            {
                continue;
            }
            else
            {
                return false;
            }
        }
        return true;
    }

    private static bool IsWithinTolerance(int currentLevel, int nextLevel)
    {
        var absDelta = Math.Abs(currentLevel - nextLevel);
        return absDelta >= 1 && absDelta <= 3;
    }
    
}

File: 2024/python/puzzles/dec02.py
from utils.file_utils import data_file_reader
from utils.output_utils import print_output
from utils.performance_utils import timing_decorator

@timing_decorator
def solve_part1(date, use_test_data = False):
	print("Day 02, Part 1: Safe reports")

	lines = data_file_reader(date = date, use_test_data = use_test_data, part = 1)
	
	total_safe_reports: int = 0

	# for each report
	for line in lines:
		report: list[str] = line.split(" ")

		if report_is_safe(report):
			total_safe_reports += 1

	print_output(str(total_safe_reports), use_test_data)


@timing_decorator
def solve_part2(date, use_test_data = False):
	print("Day 02, Part 2: Problem Dampener")

	lines = data_file_reader(date = date, use_test_data = use_test_data, part = 1)
	
	total_safe_reports: int = 0

	# for each report
	for line in lines:
		report: list[str] = line.split(" ")

		if report_is_safe_pt2(report):
			total_safe_reports += 1

	print_output(str(total_safe_reports), use_test_data)


def report_is_safe(report: list[str]) -> bool:
	if is_gradually_increasing(report) or is_gradually_decreasing(report):
		return True
	else:
		return False

def report_is_safe_pt2(report: list[str]) -> bool:
	if is_gradually_increasing(report) or is_gradually_decreasing(report):
		return True
	elif safe_with_bad_value(report):
		return True
	else:
		return False

def safe_with_bad_value(report: list[str]) -> bool:
	for i in range(len(report)):
		report_list = list(report)
		report_list.pop(i)
		if is_gradually_increasing(report_list) or is_gradually_decreasing(report_list):
			return True
	return False

def is_gradually_increasing(report: list[str]) -> bool:
	for i in range(len(report) - 1):
		current_level = int(report[i])
		next_level = int(report[i + 1])
		if current_level < next_level and is_within_tolerance(current_level, next_level):
			continue
		else:
			return False
	return True


def is_gradually_decreasing(report: list[str]) -> bool:
	for i in range(len(report) - 1):
		current_level = int(report[i])
		next_level = int(report[i + 1])
		if current_level > next_level and is_within_tolerance(current_level, next_level):
			continue
		else:
			return False
	return True


def is_within_tolerance(current_level: int, next_level: int) -> bool:
	abs_delta = abs(current_level - next_level)
	if abs_delta >= 1 and abs_delta <= 3:
		return True
	else:
		return False

